src/etc/inc/ipsec.inc

```diff
@@ -2132,6 +2136,12 @@ function ipsec_setup_tunnels() {
 			$conn['tfc_padding'] = (isset($ph1ent['tfc_bytes']) && is_numericint($ph1ent['tfc_bytes'])) ? $ph1ent['tfc_bytes'] : "mtu";
 		}
 
+		/* Custom Port */
+		if (isset($ph1ent['ikeport'])) {
+			$conn['local_port'] = isset($config['ipsec']['port_nat_t']) ? $config['ipsec']['port_nat_t'] : 4500;
```

Comment:
Thinking about this line some more, should it really depend on `ikeport`? Someone may want to change the local port but not remote.

Perhaps instead it should be its own separate test that only sets `local_port` when `port_nat_t` is present in the config without any relationship to the `ikeport` setting

Reply:
And this may need tested but do you even need to set `local_port` if you changed the value of `port_nat_t` in charon? I could be misreading the docs, since they are ambiguous, but it sounds like if you change `port_nat_t` then that would change the port automatically used by strongSwan when it selects NAT-T, so `local_port` could stay unset as it would be picked automatically.

But that needs testing to confirm.

Reply:
About setting local_port when remote_port is set, I did tests with the old ipsec.conf and it was required, at least to have a working client role. I will do some tests again with various configurations and let you know the results.

Reply:
I have set up a server listening on port 55555 (port_nat_t=55555).

Then, I tested the following configurations on client, using old ipsec.conf, with the associated results (l=local_port (or leftikeport), r=remote_port, n=port_nat_t):
- (l=unset, r=55555, n=unset) => KO, ike init sent but not accepted, client local=500 (isakmp: parent_sa ikev2_init[I])
- (l=4500, r=55555, n=unset) => working (NONESP-encap: isakmp: parent_sa ikev2_init[I])
- (l=44444, r=55555, n=unset) => KO, no init sent
- (l=0, r=55555, n=unset) => KO, something sent, client local = 500 (isakmp:)

on all, when replying the server was using source port 55555.

Then, I have set up the port_nat_t on the client-side and had the following results:

- (l=unset, r=55555, n=44444) => KO, ike init sent but not accepted, client local=500 (isakmp: parent_sa ikev2_init[I])
- (l=4500, r=55555, n=44444) => KO, no init sent
- (l=44444, r=55555, n=44444) => working (NONESP-encap: isakmp: parent_sa ikev2_init[I])
- (l=0, r=55555, n=44444) => KO, something sent, client local = 500 (isakmp:)

So imo, the conclusion is that the local port has to be set to the same value as the port_nat_t when using remote_port (or 4500 if port_nat_t is not explicitly set).

Reply:
I had a second thought on this.

When the local_port is not set, the client is connecting on the ike port, as set in remote_port.
When the local_port is set (properly), the client is connecting on the nat-t port, as set in remote_port.

So it seems to me the problem is:
- one can also want to change the ike port (on the server side)
- one can also want to connect on a non-standard ike port (from the client side)

This PR was only about the nat-t/encaps port but some more changes to handle the ike port too would make sense. In that case, I would see two ports with the remote gateway address on phase 1, one for ike, one for natt/encaps, but the user could only fill one, depending on whether he wants to connect to one or the other.

Does that sound good to you ?

Reply:
> one for ike, one for natt/encaps, but the user could only fill one, depending on whether he wants to connect to one or the other

Thinking again: both ports may be needed to have correct route-to rules.

So two ports that can be set, with priority on ike:
- if ike port is set, connect on remote ike port, regardless to nat-t port
- if nat-t port is set, connect on remote nat-t port, but only if ike port is not set